torch/sparse/triton_ops/triton_bsr_dense_mm.py

```diff
@@ -0,0 +1,295 @@
+        return torch.zeros(
+            original_batch_dims_broadcasted + (m, n),
+            dtype=dense.dtype,
+            device=dense.device,
+        )
+
+    # Introduce fake batch dimension if not present for convenience.
+    def unsqueeze_batch_dim(t, n_non_batch_dims):
+        if t.dim() > n_non_batch_dims:
+            return t
+        else:
+            return t.unsqueeze(0)
+
+    crow_indices = unsqueeze_batch_dim(bsr.crow_indices(), 1)
+    col_indices = unsqueeze_batch_dim(bsr.col_indices(), 1)
+    values = unsqueeze_batch_dim(bsr.values(), 3)
+    dense = unsqueeze_batch_dim(dense, 2)
+    nnz = values.shape[-3]
+    blocksize = values.shape[-2:]
+
+    # Compute broadcasted batch dimension
+    bsr_batch_dims = values.shape[:-3]
+    dense_batch_dims = dense.shape[:-2]
+    batch_dims_broadcasted = torch.broadcast_shapes(bsr_batch_dims, dense_batch_dims)
+
+    # Allocate output
+    output = torch.zeros(
+        batch_dims_broadcasted + (m, n), dtype=dense.dtype, device=dense.device
+    )
+
+    # Broadcast batch dimensions and squash
+    def batch_broadcast_and_squash(t, batch_dims, invariant_dims):
+        return t.broadcast_to(batch_dims + invariant_dims).flatten(
+            0, len(batch_dims) - 1
+        )
+
+    crow_indices = batch_broadcast_and_squash(
+        crow_indices, batch_dims_broadcasted, (-1,)
+    )
+    col_indices = batch_broadcast_and_squash(col_indices, batch_dims_broadcasted, (-1,))
+    values = batch_broadcast_and_squash(
+        values, batch_dims_broadcasted, values.shape[-3:]
+    )
+    dense = batch_broadcast_and_squash(dense, batch_dims_broadcasted, dense.shape[-2:])
+    # NOTE: output is contiguous, so batch_broadcast_and_squash will create a view
+    output = batch_broadcast_and_squash(
+        output, batch_dims_broadcasted, output.shape[-2:]
+    )
+
+    # NOTE: this function will ALWAYS create a view
+    def tile_to_blocksize(t, blocksize):
+        *rest, m, n = t.shape
+        new_shape = rest + [
+            m // blocksize[0],
+            blocksize[0],
+            n // blocksize[1],
+            blocksize[1],
+        ]
+        return t.reshape(new_shape).transpose(-3, -2)
+
+    # "Blockify" the row dimension of dense with blocksize[1]
+    # since dense is on the rhs of matmul
+    dense = tile_to_blocksize(dense, blocksize[::-1])
+    # "Blockify" the row dimension of output with blocksize[0]
+    # which is inherited from the bsr input.
+    # NOTE: tile_to_blocksize will create a view.
+    # NOTE: output.blocksize[-1] == dense.blocksize[-1],
+    # so it could be any value in [1, dense.shape[-1]).
+    # We need to probably use the largest possible blocksize
+    # so that it fits into SRAM.
+    output = tile_to_blocksize(output, (blocksize[0], blocksize[0]))
+
+    # Compute a vector of non-zero elements numbers per each row.
+    # We want to ultimately iterate over non-zero rows.
+    nnz_per_row = crow_indices[:, 1:] - crow_indices[:, :-1]
```

Comment:
I suppose this could be done on the fly within the kernel?

Reply:
Yes, but only if we have plenty of nonzero rows.

Reply:
I mean, with modern CUDA you can have a lot of blocks. If you exit them quickly the overhead of a large grid with lots of unused blocks might not be better than the overhead of doing these calculations first.

Reply:
Sure, makes sense, given that nonzero is  already iterating over nnzs. I will test it out and update benchmarks.

Reply:
Introduced a separate kernel for dense rowspaces. I will try to benchmark some reasonable threshold for kernel switch.